node/src/components/linear_chain.rs

```diff
@@ -134,39 +180,58 @@ where
+                if !self
+                    .linear_chain_state
+                    .add_pending_finality_signature(*fs.clone(), gossiped)
+                {
+                    // If we did not add the signature it means it's either incorrect or we already
+                    // know it.
                     return Effects::new();
                 }
-                match self.get_signatures(&block_hash) {
+                match self.linear_chain_state.get_signatures(&block_hash) {
+                    // Not found in the cache, look in the storage.
                     None => effect_builder
                         .get_signatures_from_storage(block_hash)
                         .event(move |maybe_signatures| {
-                            let maybe_box_signatures = maybe_signatures.map(Box::new);
-                            Event::GetStoredFinalitySignaturesResult(fs, maybe_box_signatures)
+                            Event::GetStoredFinalitySignaturesResult(
+                                fs,
+                                maybe_signatures.map(Box::new),
+                            )
                         }),
                     Some(signatures) => effect_builder.immediately().event(move |_| {
                         Event::GetStoredFinalitySignaturesResult(fs, Some(Box::new(signatures)))
                     }),
                 }
             }
             Event::GetStoredFinalitySignaturesResult(fs, maybe_signatures) => {
-                if let Some(signatures) = &maybe_signatures {
-                    if signatures.era_id != fs.era_id {
+                if let Some(known_signatures) = &maybe_signatures {
+                    // If the newly-received finality signature does not match the era of previously
+                    // validated signatures reject it as they can't both be
+                    // correct – block was created in a specific era so the IDs have to match.
+                    if known_signatures.era_id != fs.era_id {
                         warn!(public_key=%fs.public_key,
-                            expected=%signatures.era_id, got=%fs.era_id,
+                            expected=%known_signatures.era_id,
+                            got=%fs.era_id,
                             "finality signature with invalid era id.");
+                        self.linear_chain_state.remove_from_pending_fs(&*fs);
                         // TODO: Disconnect from the sender.
-                        self.remove_from_pending_fs(&*fs);
                         return Effects::new();
                     }
-                    // Populate cache so that next finality signatures don't have to read from the
-                    // storage. If signature is already from cache then this will be a noop.
-                    self.cache_signatures(*signatures.clone());
+                    if known_signatures.has_proof(&fs.public_key, &fs.signature) {
```

Comment:
I thought we decided that `public_key` is enough? Because it doesn't matter if the same validator created another, different, but also valid signature.

Reply:
Done in [`f5f6038` (#1268)](https://github.com/CasperLabs/casper-node/pull/1268/commits/f5f603830d33577186c4c7072cbb578714d8e773)